test/e2e/helpers.go

```diff
@@ -778,13 +784,31 @@ func getImageSkusInOffer(ctx context.Context, location, publisher, offer string)
 	return skus
 }
 
-// parseImageSkuNames parses SKU names in format "k8s-1dot17dot2-os-123" to extract the Kubernetes version.
-// it returns a sorted list of all k8s versions found.
-func parseImageSkuNames(skus []string) map[string]semver.Version {
-	capiSku := regexp.MustCompile(`^k8s-(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)-[a-z]*.*$`)
+// parseImageSkuNames parses the image SKU name and returns a map of the string version to semver.Version.
+// If the SKU name is the new format, the versions in the SKU will be parsed. If it's in the old format,
+// the version will be parsed from the SKU name, and it is assumed to contain one version.
+func parseImageSkuNames(ctx context.Context, skus []string, location, publisher, offer string) map[string]semver.Version {
+	capiSku := regexp.MustCompile(`^[\w-]+-gen[12]$`)
+	capiVersion := regexp.MustCompile(`^(\d)(\d{1,2})\.(\d{1,2})\.\d{8}$`)
+	oldCapiSku := regexp.MustCompile(`^k8s-(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)dot(0|[1-9][0-9]*)-[a-z]*.*$`)
+
 	versions := make(map[string]semver.Version, len(skus))
 	for _, sku := range skus {
 		match := capiSku.FindStringSubmatch(sku)
+		if len(match) != 0 {
+			imagesClient := newImagesClient()
+			res, err := imagesClient.List(ctx, location, publisher, offer, sku, "", nil, "")
```

Comment:
we already have to list images for every SKU earlier to make sure that they contain published images https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2388/files#diff-ead556dbd7c8058f584b092a1158049cf3350fc235fdeb762b2a4486abb610c4R775

Would it be possible to optimize it so we only list images once for each SKU?

Reply:
Yes, probably, just requires a more thorough refactoring than ~30 minutes.  😄

I'll try again, let me know if you see anything else here to improve.

Reply:
Done: I refactored the two helper funcs into one `getVersionsForOffer` func, so this PR doesn't introduce additional API calls.

Reply:
Done: I refactored the two helper funcs into one `getVersionsForOffer` func, so this PR doesn't introduce additional API calls.